File: recipes/subms-arena-allocator/rust/examples/growth_main.rs
//! Storage-growth capture for the bump arena: run many alloc-then-`reset`
//! sessions and confirm resident memory returns to the same steady level every
//! round. A bump allocator has zero per-object overhead, so resident tracks the
//! live allocations exactly (amplification ~1x), and `reset` reclaims all of it -
//! the arena never accretes across sessions.
//!
//! Emits the stable subms growth JSON on stdout.
//!
//! ```sh
//! cat <<EOF | cargo run --release --example growth_main --features harness
//! rounds=50
//! allocs_per_round=20000
//! EOF
//! ```

use std::collections::BTreeMap;
use std::io::{self, Read};
use std::process::ExitCode;

use subms::{SubMsGrowthClass, SubMsGrowthRecipe, grow, growth_to_json};
use subms_arena_allocator::Bump;

struct ArenaChurn {
    arena: Bump,
    rounds: usize,
    allocs_per_round: usize,
}

impl SubMsGrowthRecipe for ArenaChurn {
    fn name(&self) -> &str {
        "subms-arena-allocator"
    }
    fn op_name(&self) -> &str {
        "alloc"
    }
    fn rounds(&self) -> usize {
        self.rounds
    }
    fn ops_per_round(&self) -> usize {
        self.allocs_per_round
    }
    fn op(&mut self, _round: usize, i: usize) {
        // Start each round's session fresh: reset reclaims the whole buffer, then
        // we bump-allocate the round's objects into it.
        if i == 0 {
            self.arena.reset();
        }
        let _ = self.arena.alloc_copy(i as u64);
    }
    fn memory_bytes(&mut self) -> u64 {
        // Resident = bytes currently handed out of the buffer (measured at the
        // round's peak, before the next round's reset).
        self.arena.used() as u64
    }
    fn live_bytes(&mut self) -> u64 {
        // Every allocated byte is live until reset, so resident == live: a bump
        // arena wastes nothing (amplification 1x).
        self.arena.used() as u64
    }
    fn structures(&mut self) -> Vec<(String, u64)> {
        vec![("live_allocs".to_string(), self.allocs_per_round as u64)]
    }
    fn expected(&self) -> (SubMsGrowthClass, f64) {
        // Resident memory must return to the same steady level every round - it
        // must not climb, which would mean reset is leaking the buffer.
        (SubMsGrowthClass::PlateauBounded, 1.5)
    }
}

fn parse_usize(map: &BTreeMap<String, String>, key: &str, default: usize) -> usize {
    map.get(key)
        .and_then(|v| v.trim().parse().ok())
        .unwrap_or(default)
}

fn main() -> ExitCode {
    let mut raw = String::new();
    if io::stdin().read_to_string(&mut raw).is_err() {
        eprintln!("growth_main: failed to read stdin");
        return ExitCode::FAILURE;
    }
    let mut map = BTreeMap::new();
    for line in raw.lines() {
        let line = line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        if let Some((k, v)) = line.split_once('=') {
            map.insert(k.trim().to_string(), v.trim().to_string());
        }
    }
    let rounds = parse_usize(&map, "rounds", 50);
    let allocs_per_round = parse_usize(&map, "allocs_per_round", 20_000);

    // Buffer sized to hold one round's allocations (u64 + alignment) with headroom
    // - the base Bump is fixed-capacity and panics rather than growing.
    let capacity = allocs_per_round * 16 + 4096;
    let mut recipe = ArenaChurn {
        arena: Bump::with_capacity(capacity),
        rounds,
        allocs_per_round,
    };
    let report = grow(&mut recipe, "rust");

    if growth_to_json(&report, &mut io::stdout().lock()).is_err() {
        eprintln!("growth_main: failed to write json");
        return ExitCode::FAILURE;
    }
    ExitCode::SUCCESS
}

File: recipes/subms-arena-allocator/java/src/main/java/com/submillisecond/recipes/arena/GrowthMain.java
package com.submillisecond.recipes.arena;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

import com.submillisecond.perf.SubMsGrowth;
import com.submillisecond.perf.SubMsGrowthRecipe;

/**
 * Storage-growth capture for the bump arena: run many alloc-then-{@code reset}
 * sessions and confirm resident memory returns to the same steady level every
 * round. A bump allocator has zero per-object overhead, so resident tracks the
 * live allocations exactly (amplification ~1x), and {@code reset} reclaims all
 * of it - the arena never accretes across sessions.
 *
 * <p>Emits the stable subms growth JSON on stdout. Mirror of the Rust
 * {@code examples/growth_main.rs}, reading the same stdin key=value config, so
 * the two curves are directly comparable.
 *
 * <p>The bytes here do NOT diverge from the Rust port. The footprint is a
 * cursor offset into the backing buffer, not a count of objects:
 * {@code allocs_per_round} allocations of 8 bytes at alignment 8 from a cursor
 * starting at 0 leave {@code used() == 8 * allocs} in both ports (160,000 at
 * the default 20,000; measured equal round for round at 5,000 allocations
 * against the Rust example). The JVM's 16-byte array header
 * and the arena shell sit outside that accounting exactly as Rust's Vec header
 * does, so both curves report the same number.
 *
 * <pre>
 * printf 'rounds=50\nallocs_per_round=20000\n' \
 *   | java -cp target/classes:... com.submillisecond.recipes.arena.GrowthMain
 * </pre>
 */
public final class GrowthMain {

    private GrowthMain() {}

    private static final class ArenaChurn implements SubMsGrowthRecipe {
        private final BumpArena arena;
        private final ByteBuffer view;
        private final int rounds;
        private final int allocsPerRound;

        ArenaChurn(int capacity, int rounds, int allocsPerRound) {
            this.arena = new BumpArena(capacity);
            // The Rust op is alloc_copy, which allocates AND stores the value;
            // the view keeps the store in the timed path so the two ops match.
            this.view = ByteBuffer.wrap(arena.bytes()).order(ByteOrder.nativeOrder());
            this.rounds = rounds;
            this.allocsPerRound = allocsPerRound;
        }

        @Override public String name() {
            return "subms-arena-allocator";
        }

        @Override public String opName() {
            return "alloc";
        }

        @Override public int rounds() {
            return rounds;
        }

        @Override public int opsPerRound() {
            return allocsPerRound;
        }

        @Override public void op(int round, int i) {
            // Start each round's session fresh: reset reclaims the whole buffer,
            // then we bump-allocate the round's objects into it.
            if (i == 0) {
                arena.reset();
            }
            int off = arena.allocate(8, 8);
            view.putLong(off, i);
        }

        @Override public long memoryBytes() {
            // Resident = bytes currently handed out of the buffer (measured at the
            // round's peak, before the next round's reset).
            return arena.used();
        }

        @Override public long liveBytes() {
            // Every allocated byte is live until reset, so resident == live: a bump
            // arena wastes nothing (amplification 1x).
            return arena.used();
        }

        @Override public Map<String, Long> structures() {
            return Map.of("live_allocs", (long) allocsPerRound);
        }

        @Override public SubMsGrowth.GrowthClass expectedClass() {
            return SubMsGrowth.GrowthClass.PLATEAU_BOUNDED;
        }

        @Override public double expectedBound() {
            // Resident memory must return to the same steady level every round - it
            // must not climb, which would mean reset is leaking the buffer.
            return 1.5;
        }
    }

    private static int parseInt(Map<String, String> cfg, String key, int fallback) {
        String v = cfg.get(key);
        if (v == null) {
            return fallback;
        }
        try {
            return Integer.parseInt(v.trim());
        } catch (NumberFormatException e) {
            return fallback;
        }
    }

    public static void main(String[] args) throws Exception {
        Map<String, String> cfg = new HashMap<>();
        try (BufferedReader in =
                new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8))) {
            String line;
            while ((line = in.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty() || line.startsWith("#")) {
                    continue;
                }
                int eq = line.indexOf('=');
                if (eq > 0) {
                    cfg.put(line.substring(0, eq).trim(), line.substring(eq + 1).trim());
                }
            }
        }

        int rounds = parseInt(cfg, "rounds", 50);
        int allocsPerRound = parseInt(cfg, "allocs_per_round", 20_000);

        // Buffer sized to hold one round's allocations (long + alignment) with
        // headroom - the base arena is fixed-capacity and throws rather than
        // growing.
        int capacity = allocsPerRound * 16 + 4096;
        var recipe = new ArenaChurn(capacity, rounds, allocsPerRound);

        Writer out = new OutputStreamWriter(System.out, StandardCharsets.UTF_8);
        SubMsGrowth.growthToJson(SubMsGrowth.grow(recipe, "java"), out);
        out.flush();
    }
}
